compiler/src/iree/compiler/Codegen/SPIRV/SPIRVLinkExecutables.cpp

```diff
@@ -40,22 +60,108 @@ struct SPIRVLinkExecutablesPass final
     }
     sourceExecutableOps.resize(retainSize);
 
-    // Guess a module name, if needed, to make the output files readable.
-    std::string moduleName = guessModuleName(moduleOp, "spirv_module");
+    // Note that at runtime, for a particular executable, only one variant of it
+    // will be loaded. So, all variants of an executable are expected to provide
+    // the exact same set of entry points; this way we can guarantee no matter
+    // which variant is chosen, we have all entry points to call into. The same
+    // entry point in different variants may have different target requirements
+    // though.
+    //
+    // The input to the linking stage are a collection of executables, each may
+    // have multiple variants, but only ever provide one entry point. Together
+    // with the above restriction, we can link two executables if and only if
+    // their variants have the exact same set of target requirements. Under such
+    // circumstances, we can make sure for a particular target requirement
+    // (loaded as one variant during runtime), we can provide all entry points.
+
+    // Build a map from all variants' target requirements to their wrapping
+    // executable ops.
+    std::map<SmallVector<ExecutableTargetAttr, 0>,
```

Comment:
if order is important then we need a key that is deterministic (not the pointer, which will result in a different order each run) - if the order isn't important then you should just use unordered_map

Reply:
We don't care about the order for linking but it indeed can cause issues for matching in lit tests.. So I changed it to compare alphabetically. Pushed a fix to https://github.com/openxla/iree/pull/15977 but forgot to do it here. Fixed now.